Testing Automatic Grammar/Diagram generation

premkumr opened the pull request:
Automatic Generation of BNF diagrams and Grammar
-------------------------------------------------------
- Refer to [Spec doc](https://docs.google.com/document/d/1Qk3wkP0POG4aZao71WT2a6B0Xf7TZ-TqzYnhGHATQcs/edit) for full details
- Adds `ebnf` shortcode with which diagrams, grammar, and tabs can be generated dynamically
- The `ebnf` shortcode takes 2 parameters
    - `type` (default: `tabs`)
         - `tabs`: Generate the grammar + diagram tabs
         - `grammar`: Generate just the grammar
         -  `diagram`: Generate just the diagram
         -  `reference`: Generate the large reference file
    - `localrefs` : comma-separated list of rules that will be linked to the same page. These rules will be anchored as #rulename instead of grammar_diagrams#rulename as they are expected to be defined on the same page in a different diagram/grammar pair. `NOTE:` The grammar_diagrams is the reference page with all the diagrams.
- The rulenames to be processed are specified as the content of the shortcode (The text between the start and end markers) as a comma separated values, eg: 
```{{<ebnf>}}frame_clause,frame_bounds,frame_start{{</ebnf>}}```
- When the `ebnf` file of a specific version (eg `preview`, `stable`, ...) is modified, only the diagrams in that version will be regenerated
- Multiple diagrams/tabs can be added on the same page without any additional parameters. (previously they had to be suffixed with unique numbers) eg:
```
{{<ebnf>}}frame_clause,frame_bounds,frame_start{{</ebnf>}}

{{<ebnf>}}select,select_options{{</ebnf>}}
```

- The `rrdiagrams.jar` has been packaged and published to [npm repository](https://www.npmjs.com/package/yb-rrdiagram)and will be automatically downloaded and installed when Hugo is set up. 
- The diagrams will be cached internally by Hugo & will be regenerated only if the base grammar file is modified.
- Both old and new schemes can coexist. This means that the pair of grammar/diagram files generated for the previous versions (eg:`stable`, `v2.2`, ...)  can continue to be included in the `.md` files till we decide to move to the new scheme permanently.

### Developer Flow
- To add diagrams/tabs/grammar, add the rulenames with the `ebnf` shortcode (as shown below) in the `.md` files.
- After modifying the base grammar file (eg. `ysql_grammar.bnf`), Hugo needs to be restarted for diagrams to be regenerated. This is because currently there is no clear way of making the .md files be directly dependent on the base_grammar file. The .md files get rebuilt, only when the shortcode changes or when the .md files themselves change.

### Production and PR Builds
- All diagrams and reference files will be regenerated on the netlify server during the site build process. This ensures all the diagrams are generated in the same manner (as compared to multiple developers generating on different OS'es)

### Change in diagram generation flow
- No more explicit downloading of jar file
- No more manual running of the jar
- No more creation of separate grammar/diagram files.

### Tabs
Will create the grammar and diagram tabs
```
{{<ebnf localrefs="window_definition">}}
- frame_clause
- frame_bounds
- frame_start
{{</ebnf>}}
```
or 
```
{{<ebnf>}}
frame_clause,frame_bounds,frame_start
{{</ebnf>}}
```

### Grammar
Will generate only the grammar
```
{{<ebnf type="grammar">}}
- frame_clause
- frame_bounds
- frame_start
{{</ebnf>}}
```

### Diagram
Will generate only the diagram
```
{{<ebnf type="diagram">}}
- frame_clause
- frame_bounds
- frame_start
{{</ebnf>}}
```

### Reference file
The large reference file with all diagrams can be generated with 

```
{{<ebnf type="reference">}}{{</ebnf>}}
```


Preview
--------
@netlify /preview/api/ysql/exprs/window_functions/invocation-syntax-semantics

jasonyb commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,66 @@
+        </div>
+        <div id="diagram{{$suffix}}" class="tab-pane fade show active" role="tabpanel" aria-labelledby="diagram-tab">
+            {{- $bnfResponse = resources.GetRemote (printf "%s&mode=diagram" $url) -}}
+            {{- print  $bnfResponse.Content | markdownify -}}
```
```suggestion
            {{- print $bnfResponse.Content | markdownify -}}
```

premkumr replied:
done

jasonyb commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,66 @@
+{{ $statements :=  .Inner }}
+{{ $type := default "tabs" (.Get "type") }}
+{{ $statements = replaceRE " *- *" "" $statements}}
```
```suggestion
{{ $statements = replaceRE "\s*-\s*" "" $statements}}
```

premkumr replied:
done

jasonyb commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,66 @@
+{{ $type := default "tabs" (.Get "type") }}
+{{ $statements = replaceRE " *- *" "" $statements}}
+{{ $statements = replaceRE "\n+" "," $statements}}
+{{ $statements = replaceRE " +" "" $statements}}
```
```suggestion
{{ $statements = replaceRE "\s+" "" $statements}}
```
doesn't this make first case redundant?

premkumr replied:
fixed.

jasonyb commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,66 @@
+{{ $statements = replaceRE " *- *" "" $statements}}
+{{ $statements = replaceRE "\n+" "," $statements}}
+{{ $statements = replaceRE " +" "" $statements}}
+{{ $statements = replaceRE ",+" "," $statements}}
```
it may be better to not allow syntax like ",,"
```suggestion
```

premkumr replied:
Multiple whitespaces and newlines are converted to commas. This could lead to multiple commas during input cleaning. If we decide on a  strict format, this wont be needed

jasonyb commented on docs/content/preview/api/ysql/syntax_resources/ysql_grammar.ebnf:
```diff
@@ -1237,7 +1237,7 @@ expression_list = expression { ',' expression } ;
 (* And for the invocation of an aggregate function, in conjunction with the GROUP BY clause.*)
 
-select_start = 'SELECT' [ 'ALL' | 'DISTINCT' [ 'ON' ( '(' expression { ',' expression } ')' ) ] ]
+select_start = 'SELECT' [ 'ALL' | 'DISTINCT' [ 'ON' ( '(' expression { ',' expression } ')' ) ] ] \
```
this should commit as a separate change beforehand or be documented

premkumr replied:
This is removed.

jasonyb commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,66 @@
+    {{- $bnfResponse = resources.GetRemote (printf "%s&mode=reference" $url) -}}
+    {{- print $bnfResponse.Content | markdownify -}}
+{{- else -}}
+    {{- warnf "invalid type for ebnf : %s" $type }}
```
```suggestion
    {{- warnf "invalid type for ebnf: %s" $type }}
```
should it error to prevent it from ever getting missed?

premkumr replied:
switched to error.

jasonyb reviewed (commented):
could it be documented (such as code comment) more thoroughly?

> type - tabs/grammar/diagram/reference (default: tabs)

would be better if `tabs/grammar/diagram/reference` wasn't all in single backticks because it's less obvious you are listing multiple separate options

also, could each be described in more detail?

> localrefs : comma separated list of rules that will be linked to the same page

What does it mean?  A link will be created (to where?)?  How does it work if both "type" and "localrefs" are specified?

> in the regular doc.md files.

regular .md files

> When editing the base grammar file (eg. ysql_grammar.bnf). Restart Hugo.

unfinished sentence

> Will generate only the diagram\

bad backslash

premkumr commented on docs/config/_default/config.toml:
```diff
@@ -29,3 +32,6 @@ enableGitInfo = true
 [markup.goldmark.parser.attribute]
     block = true
+
+[security.http]
+mediaTypes = ['^text/plain$', 'text/plain' ]
```
Leftover from testing. Removed.

premkumr commented on docs/content/preview/api/ysql/exprs/window_functions/invocation-syntax-semantics.md:
```diff
@@ -34,85 +34,38 @@ Notice these three different orthography styles:
-  </div>
-</div>
+{{<ebnf localrefs="window_definition">}}
+
```
New lines in the list format were added as markdown linters expect a new line before lists. If we decide to go with just comma-separated values, this won't be needed.

premkumr commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,66 @@
+    {{- print $bnfResponse.Content | markdownify -}}
+{{- else -}}
+    {{- warnf "invalid type for ebnf : %s" $type }}
+{{- end -}}
```
Added

premkumr commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,66 @@
+{{ $statements :=  .Inner }}
+{{ $type := default "tabs" (.Get "type") }}
+{{ $statements = replaceRE " *- *" "" $statements}}
+{{ $statements = replaceRE "\n+" "," $statements}}
+{{ $statements = replaceRE " +" "" $statements}}
+{{ $statements = replaceRE ",+" "," $statements}}
+{{ $statements = trim $statements ","}}
+{{ $local := default "" (.Get "localrefs") }}
+{{ $path := split $.Page.File.Dir "/" }}
+{{ $version := index $path 0 }}
+{{ $depth := add 1 (len $path) }}
+{{ $api := "ysql" }}
+{{ if strings.Contains $.Page.File.Dir "/ycql/" }}{{ $api = "ycql" }}{{ end }}
+{{ $url := "http://localhost:1314/bnf" }}
+{{ $url = printf "%s?depth=%d&api=%s&version=%s&rules=%s" $url $depth $api $version $statements}}
+{{ if $local }}{{ $url = printf "%s&local=%s" $url $local }}{{ end }}
+{{ if $.Site.Params.regenerateDiagrams }}
+{{ $url = printf "%s&ts=%d" $url now.Unix }}
+{{ else }}
+{{ $f := os.Stat (printf "content/%s/api/%s/syntax_resources/%s_grammar.ebnf" $version $api $api) }}
+{{ $url = printf "%s&ts=%d" $url $f.ModTime.Unix }}
+{{ end }}
```
This is the pain of the [Go Templating Language](https://pkg.go.dev/text/template#hdr-Actions) that Hugo supports.

premkumr commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,66 @@
+        </div>
+    </div>
+{{- else if eq $type "grammar" -}}
+    {{- $bnfResponse = resources.GetRemote (printf "%s&mode=grammar" $url) -}}
```
That was the first thing that I looked into. Hugo does not let us execute external binaries.

jasonyb commented:
> The ebnf shortcode takes 2 parameters

It doesn't explain the contents between the start and end markers (whatever the technical term for those is): comma separated or bulleted list of rule names, where these names have to match ebnf file names without extension, I'm guessing.  I think this is the biggest miss in the clarity of the description.

> These rules will be anchored as #rule instead of grammar_diagrams#rule as they are expected to be defined on the same page in a different diagram/grammar pair.

Where did "grammar_diagrams" come from?  Is "rule" literally "rule" or a placeholder for a rule name, like `<rule>`?  Assuming it is a placeholder, what is the exact syntax of `<rule>`: is there any transformation of underscores to dashes, for example?

What is the alternative for rules not listed under localrefs?

> When the base ebnf file of a specific version is modified, only the diagrams in that version will be regenerated

I'm not sure what you mean.  What is "version"?  What is "base ebnf file"?

> Multiple diagrams/tabs can be added with the same shortcode without any additional parameter

Will it be one set of tabs for all rules or one set for each rule?  Also, it is not mentioned _where_/_how_ this is added.

> All diagrams and the reference file will be regenerated on the netlify server during build before being pushed to production

Is this something you changed in the code, or is it a general "you should know this"?  If the latter, it could be clearer by saying that that is the plan/expectation, which is outside of the changes made into this repository.

> The rrdiagrams.jar is now packaged and published to npm repository
>
> This jar will be automatically downloaded and installed when Hugo is set up

The first sentence is an external detail.  The way/tense it is worded could be interpreted as something your PR is enabling.  To make it better, you could use it as context like "... has been published to the npm repository, and it will be downloaded from there when Hugo is set up".

What does "installed" mean: "This jar will be automatically ... installed"?

> Both old and new schemes can coexist.

How exactly?  There are a lot of questions it makes me think of.

> regular .md files

I had a review comment before about "doc.md", but I notice it is still confusing what exactly "regular" refers to.  Something like ".md files for general pages" might be better, but @bllewell might have a better opinion.

> After modifying the base grammar file (eg. ysql_grammar.bnf), Hugo needs to be restarted for diagrams to be regenerated.

I thought that specifying the timestamp in the internal web request was supposed to eliminate this need.

premkumr commented:
I've updated the description accordingly.

> > The ebnf shortcode takes 2 parameters
> It doesn't explain the contents between the start and end markers (whatever the technical term for those is): comma separated or bulleted list of rule names, where these names have to match ebnf file names without extension, I'm guessing. I think this is the biggest miss in the clarity of the description.
- The rule names to be processed are specified as the content of the shortcode (The text between the start and end markers) as a comma separated values, eg:
```{{<ebnf>}}frame_clause,frame_bounds,frame_start{{</ebnf>}}```

 
> > These rules will be anchored as #rule instead of grammar_diagrams#rule as they are expected to be defined on the same page in a different diagram/grammar pair. 
> Where did "grammar_diagrams" come from? Is "rule" literally "rule" or a placeholder for a rule name, like `<rule>`? Assuming it is a placeholder, what is the exact syntax of `<rule>`: is there any transformation of underscores to dashes, for example?
- By `rule`, I meant rule names , I've updated the description to reflect this correctly. The `grammar_diagrams` file is the reference file that contains all the diagrams and grammar. (This exists currently)

> What is the alternative for rules not listed under localrefs?
- What do you mean by alternative? By default, the diagram has local references for rule names within the specified group. The `localrefs` option has been added to refer to other rules in a different tab pair on the same page. If not specified, any references to rules not part of the current set will be pointed to the reference file (This is the current behavior).

> > When the base ebnf file of a specific version is modified, only the diagrams in that version will be regenerated
> I'm not sure what you mean. What is "version"? What is "base ebnf file"?
- The primary grammar file eg: `ysql_grammar.ebnf`. Version is one of `preview`, `stable`, v2.12` ... 

> > Multiple diagrams/tabs can be added with the same shortcode without any additional parameter
> Will it be one set of tabs for all rules or one set for each rule? Also, it is not mentioned _where_/_how_ this is added.
- ```{{<ebnf>}}frame_clause,frame_bounds,frame_start{{</ebnf>}}

{{<ebnf>}}select,select_options{{</ebnf>}}```

> > All diagrams and the reference file will be regenerated on the netlify server during the build before being pushed to production
> Is this something you changed in the code, or is it a general "you should know this"? If the latter, it could be clearer by saying that that is the plan/expectation, which is outside of the changes made into this repository.
- Added appropriate explanation

> > The rrdiagrams.jar is now packaged and published to npm repository
> > This jar will be automatically downloaded and installed when Hugo is set up
> The first sentence is an external detail. The way/tense it is worded could be interpreted as something your PR is enabling. To make it better, you could use it as context like "... has been published to the npm repository, and it will be downloaded from there when Hugo is set up".
> What does "installed" mean: "This jar will be automatically ... installed"?
- Fixed the wording

> > Both old and new schemes can coexist.
> How exactly? There are a lot of questions it makes me think of.
- This means that the previously generated grammar/diagram file pairs can continue to be included in the md files till we completely migrate to the new scheme. This is primarily available for testing the stability and consistency of the new scheme. For eg. only 2 md files are using the on-the-fly generated diagrams while the rest of the pages still use the old scheme. Once we feel comfortable with the new scheme, we can completely move over to just the new scheme.


> > regular .md files
> I had a review comment before about "doc.md", but I notice it is still confusing what exactly "regular" refers to. Something like ".md files for general pages" might be better, but @bllewell might have a better opinion.
- I've reworded it to be just `.md` files

> > After modifying the base grammar file (eg. ysql_grammar.bnf), Hugo needs to be restarted for diagrams to be regenerated.
> I thought that specifying the timestamp in the internal web request was supposed to eliminate this need.
- This is because currently there is no clear way of making the .md files be directly dependent on the `grammar.ebnf` file. The .md files get rebuilt, only when the shortcode changes or when the .md files themselves change. As of now Hugo does not trigger rebuild when the  `grammar.ebnf`  changes (locally during development). This in a way works in our favor, as small intermittent partial saves won't trigger diagram re-generation.

jasonyb commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,94 @@
+{{/* Get the inner content of the shortcode */}}
+{{ $statements :=  .Inner }}
+{{/* type has to be one of [tabs, grammar, diagram reference] (default:tabs) */}}
```
```suggestion
{{/* type has to be one of [tabs, grammar, diagram, reference] (default:tabs) */}}
```

premkumr replied:
fixed

jasonyb commented on docs/package.json:
```diff
@@ -4,16 +4,20 @@
     "hugo-server-production": "hugo server --bind 0.0.0.0 --baseURL ${YB_HUGO_BASE:-localhost} --environment production",
     "hugo-server-static": "hugo server --renderStaticToDisk --bind 0.0.0.0 --baseURL ${YB_HUGO_BASE:-localhost}",
+    "start-diagrams": "npm install --no-audit --no-fund && java -jar node_modules/yb-rrdiagram/rrdiagram*.jar --server --debug",
+    "stop-diagrams": "curl --silent localhost:1314/shutdown",
```
```suggestion
    "stop-diagrams": "curl --silent http://localhost:1314/shutdown",
```

probably better this way

premkumr replied:
added

jasonyb commented on docs/layouts/shortcodes/ebnf.html:
```diff
@@ -0,0 +1,94 @@
+    {{- print $bnfResponse.Content | markdownify -}}
+
+{{- else -}}
+    {{- errorf "invalid type for ebnf : %s" $type }}
```
```suggestion
    {{- errorf "invalid type for ebnf: %s" $type }}
```

premkumr replied:
fixed

jasonyb reviewed (approved):
seems good

aishwarya24 reviewed (approved):
LGTM. This is huge! Thanks @premkumr. Tested the set up locally and it works as expected.

bllewell commented:
I'll approve it now, I focussed only on the Spec document. I've asked several times that you use "EBNF: thoughout and never just plain "BNF". Pls fix this now.

bllewell reviewed (approved):
LGTM
